internal/cluster/discovery_test.go

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (c) 2008-2023, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cluster_test
+
+import (
+	"context"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	pubcluster "github.com/hazelcast/hazelcast-go-client/cluster"
+	"github.com/hazelcast/hazelcast-go-client/cluster/discovery"
+	"github.com/hazelcast/hazelcast-go-client/internal/cluster"
+	"github.com/hazelcast/hazelcast-go-client/internal/logger"
+)
+
+type dummyDiscoverStrategy struct {
+	nodes []discovery.Node
+}
+
+func (d dummyDiscoverStrategy) DiscoverNodes(ctx context.Context) ([]discovery.Node, error) {
+	return d.nodes, nil
+}
+
+func TestDiscoveryStrategyAdapter_Addresses(t *testing.T) {
+	testCases := []struct {
+		name     string
+		nodes    []discovery.Node
+		target   []pubcluster.Address
+		publicIP bool
+	}{
+		{
+			name: "no public or private addreses",
+			nodes: []discovery.Node{
+				{},
+			},
+			target: []pubcluster.Address{},
+		},
+		{
+			name: "no public ip with public and private adresses",
+			nodes: []discovery.Node{
+				{PrivateAddr: "10.10.10.10:5000", PublicAddr: "20.10.10.10:5000"},
+				{PrivateAddr: "10.11.11.11:5000", PublicAddr: "20.11.11.11:5000"},
+			},
+			target: []pubcluster.Address{
+				"10.10.10.10:5000",
+				"10.11.11.11:5000",
+			},
+		},
+		{
+			name: "no public ip with only private adresses",
+			nodes: []discovery.Node{
+				{PrivateAddr: "10.10.10.10:5000"},
+				{PrivateAddr: "10.11.11.11:5000"},
+			},
+			target: []pubcluster.Address{
+				"10.10.10.10:5000",
+				"10.11.11.11:5000",
+			},
+		},
+		{
+			name: "no public ip with only public adresses",
+			nodes: []discovery.Node{
+				{PublicAddr: "20.10.10.10:5000"},
+				{PublicAddr: "20.11.11.11:5000"},
+			},
+			target: []pubcluster.Address{
+				"20.10.10.10:5000",
+				"20.11.11.11:5000",
+			},
+		},
+		{
+			name:     "public ip with public and private adresses",
+			publicIP: true,
+			nodes: []discovery.Node{
+				{PrivateAddr: "10.10.10.10:5000", PublicAddr: "20.10.10.10:5000"},
+				{PrivateAddr: "10.11.11.11:5000", PublicAddr: "20.11.11.11:5000"},
+			},
+			target: []pubcluster.Address{
+				"20.10.10.10:5000",
+				"20.11.11.11:5000",
+			},
+		},
+		{
+			name:     "public ip with only private adresses",
```

Comment:
This case throws in java, if you request public IP but it cannot be translated the client throws. See my other comments, I mentioned about this.

Reply:
The idea to allow that is, if the discovery strategy doesn't care about `usePublicIP` setting, it would just return the addresses it has, and the Go client would remedy that by copying them to public addresses if  `usePublicIP == true`.

Reply:
If usePublicIP == true, why do we use private IP by copying it to public IP? I think go client should respect the usePublicIp config. In java, we throw as far as I see. I did not get why we do differently in go.

Reply:
Go client respects the `usePublicIP` config if both publicIP and privateIP exists. If one of them is missing, Go client tries to remedy that instead of panicking.